kofserver/scoreboard.py

```diff
@@ -97,6 +121,71 @@ def LogPlayerAction(gameName, playerName, actionType, actionContent):
     # [ { "playerName": "Alice", "portUptime": 0, "portScore": 0,
     #     "pidUptime": 60, "pidScore": 300, "totalScore": 300 } ]
-    def QueryScore(gameName, playerName):
-        # TODO
-        return []
+    def QueryScore(self, gameName, playerName):
+        sql = "SELECT player_name, SUM(port_uptime), SUM(port_score_per_sec), SUM(pid_uptime), SUM(pid_score_per_sec) FROM score %s GROUP BY game_name, player_name;"
+        whereSQL = ""
+        if gameName != "":
+            whereSQL += "WHERE game_name='%s'" % (gameName)
+        if playerName != "":
+            whereSQL += "WHERE player_name='%s'" % (playerName) if whereSQL == "" else " AND player_name='%s'" % (playerName)
+        
+        sql = sql % (whereSQL)
+        cursor = self.database.Execute(sql)
+        try:
+            records = cursor.fetchall()
+            cursor.close()
+            results = []
+            if len(records) > 0:
+                for r in records:
+                    portScore = r[1] * r[2]
+                    pidScore = r[3] * r[4]
+                    totalScore = portScore + pidScore
+                    # If playerName is "", then all players with non-zero scores are returned.
+                    if playerName == "" and totalScore == 0:
+                        continue
+                    results.append({
+                        "playerName": r[0],
+                        "portUptime": r[1],
+                        "portScore": portScore,
+                        "pidUptime": r[3],
+                        "pidScore": pidScore,
+                        "totalScore": totalScore})
+            return results
+        except:  
+            raise
+
+    def CreateScoreTable(self):
+        sql = """CREATE TABLE IF NOT EXISTS score
+                (id INTEGER PRIMARY KEY AUTOINCREMENT,
+                game_name TEXT NOT NULL,
+                player_name TEXT NOT NULL,
+                port_uptime INTEGER,
+                port_score_per_sec INTEGER,
+                pid_uptime INTEGER,
+                pid_score_per_sec INTEGER,
+                create_time INTEGER NOT NULL);"""
+        cursor = self.database.Execute(sql)
+        try:
+            self.database.Commit()
+            results = False if cursor.rowcount == -1 else True
+            cursor.close()
+            return results
+        except:  
+            raise
+
+    def CreateActionTable(self):
+        sql = """CREATE TABLE IF NOT EXISTS action
+                (id INTEGER PRIMARY KEY AUTOINCREMENT,
+                game_name TEXT NOT NULL,
+                player_name TEXT NOT NULL,
+                type TEXT NOT NULL,
+                content TEXT,
+                create_time INTEGER NOT NULL);"""
+        cursor = self.database.Execute(sql)
+        try:
+            self.database.Commit()
+            results = False if cursor.rowcount == -1 else True
+            cursor.close()
+            return results
+        except:  
+            raise
```

Comment:
Nit/Optional: Please add new line at the end of the file.

Reply:
Done